truffle/src/main/java/org/jruby/truffle/nodes/core/ArrayNodes.java

```diff
@@ -1876,20 +1944,35 @@ public RubyArray mapObject(VirtualFrame frame, RubyArray array, RubyProc block)
         }
     }
 
-    @CoreMethod(names = {"map!", "collect!"}, needsBlock = true, raiseIfFrozenSelf = true)
+    @CoreMethod(names = {"map!", "collect!"}, needsBlock = true)
     @ImportGuards(ArrayGuards.class)
     public abstract static class MapInPlaceNode extends YieldingCoreMethodNode {
 
-        @Child private ArrayBuilderNode arrayBuilder;
+        @Child private CallDispatchHeadNode toEnumNode;
+        @Child private ArrayWriteDenormalizedNode writeNode;
+
+        private final BranchProfile breakProfile = BranchProfile.create();
+        private final BranchProfile nextProfile = BranchProfile.create();
+        private final BranchProfile redoProfile = BranchProfile.create();
 
         public MapInPlaceNode(RubyContext context, SourceSection sourceSection) {
             super(context, sourceSection);
-            arrayBuilder = new ArrayBuilderNode.UninitializedArrayBuilderNode(context);
         }
 
         public MapInPlaceNode(MapInPlaceNode prev) {
             super(prev);
-            arrayBuilder = prev.arrayBuilder;
+            toEnumNode = prev.toEnumNode;
+            writeNode = prev.writeNode;
+        }
+
+        @Specialization
+        public Object eachEnumerator(VirtualFrame frame, RubyArray array, UndefinedPlaceholder block) {
+            if (toEnumNode == null) {
+                CompilerDirectives.transferToInterpreter();
+                toEnumNode = insert(DispatchHeadNodeFactory.createMethodCall(getContext()));
+            }
+
+            return toEnumNode.call(frame, array, "to_enum", null, getContext().getCoreLibrary().getMapBangSymbol());
```

Comment:
I see 90 occurrences of `returns an Enumerator` in the spec/ruby directory so I agree it would probably make a good annotation. I think this should be a different PR though.